camayoc/qcs_models.py

```diff
@@ -93,50 +213,74 @@ def __eq__(self, other):
+        >>> hostcred.create()
+        >>> netprof = NetworkProfile(hosts=['0.0.0.0'],
+                                     credential_ids=[hostcred._id])
+        >>> netprof.create()
+        >>> actual_prof = netprof.read_self().json()
+        >>>
+        >>> # the __eq__() method is customized to
+        >>> # allow == comparison between the json returned by the server
+        >>> # and the NetworkProfile object used to create it.
+        >>>
+        >>> assert actual_prof == netprof
+    """
 
     def __init__(
             self,
+            client=None,
             name=None,
             hosts=None,
             ssh_port=None,
             credential_ids=None,
             _id=None):
-        """Objects of this type contain data associated with network profiles.
-
-        Network profiles can be created on the quipucords server by
-        instantiating a NetworkProfile object. A unique name and username are
-        provided by default. In order to create a valid network profile,
-        you must specify at least one existing host credential and one host.
+        """Iniitalize a NetworkProfile object with given data.
 
-        Example::
-            >>> from camayoc import api
-            >>> from camayoc.qcs_models import NetworkProfile
-            >>> client = api.QCSClient()
-
-            >>> hostcred = HostCredential(password='foo')
-            >>> create_response = client.create_host_cred(hostcred.payload())
-            >>> hostcred._id = create_response.json()['id']
-
-            >>> netprof = NetworkProfile(hosts=['0.0.0.0'],
-                                         credential_ids=['host._id'])
-            >>> client.create_net_prof(netprof.payload())
+        If no ssh_port is supplied, it will be set to 22 by default.
+        A uuid4 name and api.Client are also supplied if none are provided.
         """
-        super().__init__(name=name, _id=_id)
+        super().__init__(client=client, name=name, _id=_id)
+        self.endpoint = QCS_PROFILES_PATH
         self.hosts = hosts
-        self.ssh_port = 22 if not name else name
+        self.ssh_port = 22 if not ssh_port else ssh_port
         self.credentials = credential_ids
 
     def __eq__(self, other):
-        """Return true if both objects are equal."""
+        """Return true if both objects are equal.
+
+        :param other: This can be either another NetworkProfile or a dictionary
+            or json object returned from the QCS server (or crafted by hand.)
+            If `other` is a NetworkProfile instance, the two object's fields()
+            will be compared. Otherwise, we must extract the credential id's
+            from the json returned by the server into a list, because that is
+            all the data that we use to generate the network profile.
+        """
         if isinstance(other, NetworkProfile):
-            return self.__dict__ == other.__dict__
+            return self.fields() == other.fields()
```

Comment:
Better to raise a value type error when the other is not a NetworkProfile object

Reply:
The intent of overriding the `__eq__()` method was so when you read the info from the server and get back json, you could say 
```python
assert json_from_server == network_profile_obj_used_to_create_it
```

Reply:
I see so maybe a comment before doing the logic below  would help understand what is going on. As I understood you are trying to:

1) Check other is false (empty, none for example) and return False
2) Check it is an instance of NetworkProfile and check if the return of fields are the same
3) Try to perform a dictionary match logic to see if they match

That said, I think you could reorganize this on that order and before the third step provide a comment stating that you are assuming the other is a dictionary and will try to match the key and values.

But there is an issue still: doing that will make it work when doing networkprofile_object == dict work but the other way is not possible dict == networkprofile_object. Internal python objects works in the both ways:

 ```python
>>> (1 == {1: 2}) == ({1: 2} == 1)
True
>>> 1 == {1: 2}
False
>>> {1: 2} == 1
False
```

That is why I think it may not be a good idea to do that.

Reply:
Ah, maybe I could put this logic in a separate `equivalent()` method so I could call `netprof.equivalent(json_dict)`

Reply:
Yeah a method would make more sense I think.

Reply:
Now I moved this logic to a method called `equivalent()` and raise a `TypeError` if it is not of the same class or a dictionary.